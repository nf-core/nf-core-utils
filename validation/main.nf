#!/usr/bin/env nextflow

// Import plugin functions
include { 
    sayHello;
    getWorkflowVersion;
    checkCondaChannels;
    completionSummary;
    generateVersionReport;
    generateCitationReport;
    processMixedVersionSources;
    processMixedCitationSources;
} from 'plugin/nf-core-utils'

workflow {
    
    log.info """
    ================================
    nf-utils Plugin Integration Test
    ================================
    """
    
    // Test 1: Basic sayHello function
    log.info "Testing sayHello function..."
    sayHello("nf-utils integration test")
    
    // Test 2: Get workflow version
    log.info "Testing getWorkflowVersion function..."
    version = getWorkflowVersion()
    log.info "Workflow version: ${version}"
    
    // Test 3: Check conda channels
    log.info "Testing checkCondaChannels function..."
    conda_check = checkCondaChannels()
    log.info "Conda channels check: ${conda_check}"
    
    // Test 4: Test version processing with sample data
    log.info "Testing version processing..."
    
    // Sample topic channel data: [process, name, version]
    sample_versions = [
        ['FASTQC', 'fastqc', '0.11.9'],
        ['MULTIQC', 'multiqc', '1.12'],
        ['TRIMGALORE', 'trimgalore', '0.6.7'],
        ['TRIMGALORE', 'cutadapt', '4.1']
    ]
    
    // Test version report generation
    version_report = generateVersionReport(sample_versions, [])
    log.info "Version report generated successfully"
    
    // Test 5: Test citation processing with sample data
    log.info "Testing citation processing..."
    
    // Create sample meta.yml files for testing
    sample_meta_paths = []
    
    // Test mixed version sources (topic + legacy)
    log.info "Testing mixed version sources..."
    mixed_versions = processMixedVersionSources(sample_versions, [])
    log.info "Mixed versions processed successfully"
    
    // Test 6: Completion summary
    log.info "Testing completion summary..."
    completionSummary(false)
    
    log.info """
    ================================
    All integration tests completed!
    ================================
    """
}
