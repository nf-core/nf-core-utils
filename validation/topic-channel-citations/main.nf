#!/usr/bin/env nextflow

/*
 * E2E validation test for nf-core-utils plugin topic channel citation management
 * 
 * This test demonstrates the new runtime citation collection approach using topic channels.
 * Instead of manually configuring citations, processes automatically emit citation data
 * which is collected and processed automatically.
 */

// Import the new auto-citation functions from plugin
include { 
    getCitation;
    autoToolCitationText;
    autoToolBibliographyText;
} from 'plugin/nf-core-utils'

nextflow.enable.dsl = 2

// Mock processes that simulate nf-core modules with topic channel citation emission
process FASTQC {
    input:
    val sample_id
    
    output:
    path "*.html", emit: html
    path "*.zip", emit: zip
    val citation_data, topic: citation
    
    script:
    citation_data = getCitation("${projectDir}/mock-data/fastqc_meta.yml")
    """
    echo "<html>FastQC Report for ${sample_id}</html>" > ${sample_id}_fastqc.html
    echo "FastQC zip data" > ${sample_id}_fastqc.zip
    """
}

process MULTIQC {
    input:
    path fastqc_files
    
    output:
    path "*.html", emit: report
    val citation_data, topic: citation
    
    script:
    citation_data = getCitation("${projectDir}/mock-data/multiqc_meta.yml")
    """
    echo "<html>MultiQC Report</html>" > multiqc_report.html
    """
}

process SAMTOOLS_VIEW {
    input:
    val sample_id
    
    output:
    path "*.sam", emit: sam
    val citation_data, topic: citation
    
    script:
    citation_data = getCitation("${projectDir}/mock-data/samtools_meta.yml")
    """
    echo "SAM data for ${sample_id}" > ${sample_id}.sam
    """
}

// Optional process that may or may not run based on parameters
process OPTIONAL_TOOL {
    when:
    params.run_optional == true
    
    input:
    val sample_id
    
    output:
    path "*.txt", emit: result
    val citation_data, topic: citation
    
    script:
    citation_data = getCitation("${projectDir}/mock-data/star_meta.yml")
    """
    echo "Optional tool output for ${sample_id}" > ${sample_id}_optional.txt
    """
}

workflow {
    
    log.info """
    ========================================================
    nf-core-utils Plugin Topic Channel Citation E2E Test
    ========================================================
    
    This test demonstrates automatic citation collection using:
    1. Runtime citation extraction via getCitation() 
    2. Topic channel emission from processes
    3. Automatic aggregation and formatting
    4. Zero-maintenance citation management
    """
    
    // Create sample data
    samples = Channel.of('sample1', 'sample2', 'sample3')
    
    // Run processes - each automatically emits citations to topic channel
    fastqc_out = FASTQC(samples)
    samtools_out = SAMTOOLS_VIEW(samples)
    
    // MultiQC processes the fastqc outputs
    multiqc_out = MULTIQC(fastqc_out.html.collect())
    
    // Optional tool runs only if parameter is set
    optional_out = OPTIONAL_TOOL(samples.first())
    
    // Collect all citations from topic channel automatically
    ch_citations = channel.topic('citation')
        .collect()
        .map { citations -> 
            log.info "=== Collected ${citations.size()} citation emissions ==="
            citations.each { citation ->
                log.info "Citation: ${citation}"
            }
            return citations
        }
    
    // Generate final citation text and bibliography automatically
    final_citations = ch_citations
        .map { citations ->
            [
                citation_text: autoToolCitationText(citations),
                bibliography: autoToolBibliographyText(citations)
            ]
        }
    
    // Write outputs and demonstrate usage
    final_citations.subscribe { citations ->
        log.info """
        ========================================
        AUTOMATIC CITATION RESULTS
        ========================================
        
        Citation Text:
        ${citations.citation_text}
        
        Bibliography:
        ${citations.bibliography}
        
        ========================================
        """
        
        // Write to files for validation
        def output_dir = "${workflow.workDir}/pipeline_info"
        
        // Create output directory
        new File(output_dir).mkdirs()
        
        // Write citation text
        new File("${output_dir}/auto_citations.txt").text = citations.citation_text
        
        // Write bibliography  
        new File("${output_dir}/auto_bibliography.html").text = citations.bibliography
        
        log.info "âœ… Auto-generated citations written to: ${output_dir}/"
        log.info "âœ… Citation text: auto_citations.txt"
        log.info "âœ… Bibliography: auto_bibliography.html"
    }
}

workflow.onComplete {
    log.info """
    ========================================================
    Topic Channel Citation Management Test Complete! 
    ========================================================
    
    âœ… Processes automatically emitted citations to topic channel
    âœ… Citations collected and processed without manual configuration
    âœ… Final citation text and bibliography generated automatically
    âœ… Zero maintenance required - citations reflect actual tool usage
    
    Key Benefits Demonstrated:
    â€¢ No manual citation configuration needed
    â€¢ Only executed tools appear in citations  
    â€¢ Automatic formatting and aggregation
    â€¢ Clean separation of concerns
    â€¢ Runtime accuracy
    
    The future of nf-core citation management! ðŸš€
    ========================================================
    """
}