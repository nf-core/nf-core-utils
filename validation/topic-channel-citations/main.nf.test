nextflow_pipeline {

    name "Test nf-core-utils topic channel citations functions"
    script "../main.nf"
    config "../nextflow.config"

    test("Save citations outputs") {

        when {
            params {
                validate_params = false
                outdir = "$outputDir"
            }
        }

        then {
            assert workflow.success
            assert snapshot(
                filterNextflowOutput(workflow.stdout + workflow.stderr, keepAnsi: true),
                getAllFilesFromDir(params.outdir)
            ).match()
        }
    }
}
