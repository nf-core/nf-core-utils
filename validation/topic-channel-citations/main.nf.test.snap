{
    "Save citations outputs": {
        "content": [
            [
                "N E X T F L O W  ~  version [VERSION]",
                "Launching `[PATH]/topic-channel-citations/../topic-channel-citations/main.nf` [RUN_NAME] DSL2 - revision: [REVISION]",
                "Pipeline is starting! \ud83d\ude80",
                "==========================================",
                "nf-core-utils Plugin Topic Channel Citation E2E Test",
                "==========================================",
                "This test demonstrates automatic citation collection using:",
                "1. Runtime citation extraction via getCitation()",
                "2. Topic channel emission from processes",
                "3. Automatic aggregation and formatting",
                "4. Zero-maintenance citation management",
                "=== Collected 7 citation emissions ===",
                "Citation: [VIEW, samtools, [description:Tools for dealing with SAM, BAM and CRAM files, homepage:http://www.htslib.org/, documentation:http://www.htslib.org/doc/samtools.html, tool_dev_url:https://github.com/samtools/samtools, doi:10.1093/bioinformatics/btp352, licence:MIT]]",
                "Citation: [FASTQC, fastqc, [description:A quality control tool for high throughput sequence data., homepage:https://www.bioinformatics.babraham.ac.uk/projects/fastqc/, documentation:https://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/, tool_dev_url:https://www.bioinformatics.babraham.ac.uk/projects/fastqc/, doi:10.1093/bioinformatics/btw354, licence:GPL v3]]",
                "Citation: [MULTIQC, multiqc, [description:MultiQC searches a given directory for analysis logs and compiles a HTML report., homepage:https://multiqc.info/, documentation:https://multiqc.info/docs/, tool_dev_url:https://github.com/ewels/MultiQC, doi:10.1093/bioinformatics/btw354, licence:GPL v3]]",
                "==========================================",
                "AUTOMATIC CITATION RESULTS",
                "==========================================",
                "Citation Text:",
                "Tools used in the workflow included: samtools (DOI: 10.1093/bioinformatics/btp352), fastqc (DOI: 10.1093/bioinformatics/btw354), multiqc (DOI: 10.1093/bioinformatics/btw354).",
                "Bibliography:",
                "<li>samtools. doi: 10.1093/bioinformatics/btp352. <a href='http://www.htslib.org/'>http://www.htslib.org/</a></li> <li>fastqc. doi: 10.1093/bioinformatics/btw354. <a href='https://www.bioinformatics.babraham.ac.uk/projects/fastqc/'>https://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></li> <li>multiqc. doi: 10.1093/bioinformatics/btw354. <a href='https://multiqc.info/'>https://multiqc.info/</a></li>",
                "==========================================",
                "\u2705 Auto-generated citations written to: [PATH]/tests/[NFT_HASH]/output/",
                "\u2705 Citation text: [PATH]/tests/[NFT_HASH]/output/auto_citations.txt",
                "\u2705 Bibliography: [PATH]/tests/[NFT_HASH]/output/auto_bibliography.html",
                "==========================================",
                "Topic Channel Citation Management Test Complete!",
                "==========================================",
                "\u2705 Processes automatically emitted citations to topic channel",
                "\u2705 Citations collected and processed without manual configuration",
                "\u2705 Final citation text and bibliography generated automatically",
                "\u2705 Zero maintenance required - citations reflect actual tool usage",
                "Key Benefits Demonstrated:",
                "\u2022 No manual citation configuration needed",
                "\u2022 Only executed tools appear in citations",
                "\u2022 Automatic formatting and aggregation",
                "\u2022 Clean separation of concerns",
                "\u2022 Runtime accuracy",
                "The future of nf-core citation management! \ud83d\ude80",
                "==========================================",
                "Pipeline complete! \ud83d\udc4b",
                "[NXF_HASH] Submitted process > FASTQC (sample1)",
                "[NXF_HASH] Submitted process > FASTQC (sample2)",
                "[NXF_HASH] Submitted process > FASTQC (sample3)",
                "[NXF_HASH] Submitted process > MULTIQC",
                "[NXF_HASH] Submitted process > SAMTOOLS_VIEW (sample1)",
                "[NXF_HASH] Submitted process > SAMTOOLS_VIEW (sample2)",
                "[NXF_HASH] Submitted process > SAMTOOLS_VIEW (sample3)"
            ],
            [
                "auto_bibliography.html:md5,f36cf075e91a7ea08b184370db54b239",
                "auto_citations.txt:md5,b6f87ba7390abff3a0bc0da48443ab7b"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.6"
        },
        "timestamp": "2025-09-08T18:19:02.521834523"
    }
}